File: 5 graphs_algorithms/1_dijkstra/dijkstra.cs
.using System;
using System.Collections.Generic;

class Dijkstra
{
    public Dictionary<char, int> ShortestDistances(Dictionary<char, Dictionary<char, int>> graph, char start)
    {
        Dictionary<char, int> distances = new Dictionary<char, int>();
        foreach (char node in graph.Keys)
        {
            distances[node] = int.MaxValue; // Initialize distances to infinity
        }
        distances[start] = 0; // Distance from start to itself is 0

        SortedSet<(int, char)> priorityQueue = new SortedSet<(int, char)>();
        priorityQueue.Add((0, start));

        while (priorityQueue.Count > 0)
        {
            (int currentDistance, char currentNode) = priorityQueue.Min;
            priorityQueue.Remove(priorityQueue.Min);

            // If current distance is greater than the known shortest distance, skip
            if (currentDistance > distances[currentNode])
            {
                continue;
            }

            // Update distances to neighbors through current node
            foreach (var neighbor in graph[currentNode])
            {
                char neighborNode = neighbor.Key;
                int weight = neighbor.Value;
                int distance = currentDistance + weight;
                if (distance < distances[neighborNode])
                {
                    distances[neighborNode] = distance;
                    priorityQueue.Add((distance, neighborNode));
                }
            }
        }

        return distances;
    }
}

class Program
{
    static void Main(string[] args)
    {
        Dictionary<char, Dictionary<char, int>> graph = new Dictionary<char, Dictionary<char, int>>
        {
            {'A', new Dictionary<char, int> { {'B', 3}, {'C', 2} }},
            {'B', new Dictionary<char, int> { {'A', 3}, {'C', 1}, {'D', 6} }},
            {'C', new Dictionary<char, int> { {'A', 2}, {'B', 1}, {'D', 4} }},
            {'D', new Dictionary<char, int> { {'B', 6}, {'C', 4} }}
        };

        Dijkstra dijkstra = new Dijkstra();
        char startNode = 'A';
        Dictionary<char, int> distances = dijkstra.ShortestDistances(graph, startNode);

        // Print shortest distances from start node to all other nodes
        foreach (var node in distances)
        {
            Console.WriteLine($"Shortest distance from {startNode} to {node.Key}: {node.Value}");
        }
    }
}

File: 5 graphs_algorithms/1_dijkstra/dijkstra.py
import heapq

def dijkstra(graph, start):
    # Initialize distances dictionary to store shortest distances from start to all nodes
    distances = {node: float('inf') for node in graph}
    distances[start] = 0  # Distance from start to itself is 0

    # Priority queue to store (distance, node) pairs
    priority_queue = [(0, start)]

    while priority_queue:
        current_distance, current_node = heapq.heappop(priority_queue)

        # If current distance is greater than the known shortest distance, skip
        if current_distance > distances[current_node]:
            continue

        # Update distances to neighbors through current node
        for neighbor, weight in graph[current_node].items():
            distance = current_distance + weight
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                heapq.heappush(priority_queue, (distance, neighbor))

    return distances

# Example graph
graph = {
    'A': {'B': 3, 'C': 2},
    'B': {'A': 3, 'C': 1, 'D': 6},
    'C': {'A': 2, 'B': 1, 'D': 4},
    'D': {'B': 6, 'C': 4}
}

start_node = 'A'
distances = dijkstra(graph, start_node)

# Print shortest distances from start node to all other nodes
for node, distance in distances.items():
    print(f"Shortest distance from {start_node} to {node}: {distance}")
